nav2_dwb_controller/dwb_controller/src/dwb_controller.cpp

```diff
@@ -127,64 +122,72 @@ DwbController::on_shutdown(const rclcpp_lifecycle::State &)
+  RCLCPP_INFO(get_logger(), "Received a goal, begin following path");
+  auto result = std::make_shared<nav2_msgs::action::FollowPath::Result>();
+
+  std::shared_ptr<GoalHandle> current_goal_handle = goal_handle;
+
+  rclcpp::Rate loop_rate(100ms);    // period vs. hz
+
+  auto goal = current_goal_handle->get_goal();
+
   try {
-    auto path = nav_2d_utils::pathToPath2D(*command);
+    auto path = nav_2d_utils::pathToPath2D(goal->path);
 
+    RCLCPP_DEBUG(get_logger(), "Providing path to the local planner");
     planner_->setPlan(path);
-    RCLCPP_INFO(get_logger(), "Initialized");
 
-    rclcpp::Rate loop_rate(10);
     while (rclcpp::ok()) {
       nav_2d_msgs::msg::Pose2DStamped pose2d;
       if (!getRobotPose(pose2d)) {
-        RCLCPP_INFO(get_logger(), "No pose. Stopping robot");
+        RCLCPP_INFO(get_logger(), "No pose. Stopping the robot");
         publishZeroVelocity();
       } else {
         if (isGoalReached(pose2d)) {
+          RCLCPP_INFO(get_logger(), "Reached the goal");
           break;
         }
+
         auto velocity = odom_sub_->getTwist();
         auto cmd_vel_2d = planner_->computeVelocityCommands(pose2d, velocity);
-        publishVelocity(cmd_vel_2d);
+
         RCLCPP_INFO(get_logger(), "Publishing velocity at time %.2f", now().seconds());
+        publishVelocity(cmd_vel_2d);
 
-        // Check if this task has been canceled
-        if (task_server_->cancelRequested()) {
-          RCLCPP_INFO(this->get_logger(), "execute: task has been canceled");
-          task_server_->setCanceled();
+        if (current_goal_handle->is_canceling()) {
+          RCLCPP_INFO(this->get_logger(), "Canceling execution of the local planner");
+          current_goal_handle->canceled(result);
           publishZeroVelocity();
-          return TaskStatus::CANCELED;
+          return;
         }
 
         // Check if there is an update to the path to follow
-        if (task_server_->updateRequested()) {
-          // Get the new, updated path
-          auto path_cmd = std::make_shared<nav2_tasks::FollowPathCommand>();
-          task_server_->getCommandUpdate(path_cmd);
-          task_server_->setUpdated();
-
-          // and pass it to the local planner
-          auto path = nav_2d_utils::pathToPath2D(*path_cmd);
+        if (action_server_->preempt_requested()) {
+          RCLCPP_INFO(get_logger(), "Received a new goal, pre-empting the old one");
+          current_goal_handle = action_server_->get_updated_goal_handle();
+          goal = current_goal_handle->get_goal();
+
+          // If so, pass the new path to the local planner
+          auto path = nav_2d_utils::pathToPath2D(goal->path);
           planner_->setPlan(path);
         }
       }
+
       loop_rate.sleep();
     }
   } catch (nav_core2::PlannerException & e) {
     RCLCPP_INFO(this->get_logger(), e.what());
     publishZeroVelocity();
-    return TaskStatus::FAILED;
+    current_goal_handle->abort(result);
```

Comment:
With lifecycle nodes, do have to do a state transition (to error or unconfigured) in the event of an exception?

Reply:
@orduno Good question. In this case, this is a node that is providing an action server. If the particular action fails with an exception and we catch it, I think it makes sense to remain running to service future action requests. We could, perhaps release the DWBLocalPlanner and create a fresh instance. We should discuss as we develop an approach for error recovery.